source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+empty batch. For these cursors, drivers support both the ``timeoutMS`` and
+``maxAwaitTimeMS`` options. The ``timeoutMS`` option is used to derive
+client-side timeouts, while the ``maxAwaitTimeMS`` option is used as the
+``maxTimeMS`` field for ``getMore`` commands. These values have distinct
+meanings, so supporting both yields a more robust, albeit verbose, API.
+Drivers error if ``maxAwaitTimeMS`` is greater than or equal to ``timeoutMS``
+because in that case, ``getMore`` requests would not succeed if the batch was
+empty: the server would wait for ``maxAwaitTimeMS``, but the driver would
+close the socket after ``timeoutMS``.
+
+Change stream behavior
+----------------------
+
+Change streams internally behave as tailable awaitData cursors, so the
+behavior of the ``timeoutMS`` option is the same for both. The main
+difference is that change streams are resumable and drivers automatically
+perform resume attempts when they encounter transient errors. This allows
+change streams to be resilient to timeouts. If ``timeoutMS`` expires during a
+next call, drivers can’t auto-resume, but they can make sure the change
+stream is not invalidated so the user can call next again. In this case, the
+subsequent call would perform the resume without doing a ``getMore`` first.
+
+withTransaction communicates ``timeoutMS`` via ClientSession
+--------------------------------------------------------
+
+Because the ``withTransaction`` API doesn’t allow drivers to plumb down the
+remaining timeout into the user-provided callback, this spec requires the
+remaining timeout to be stored on the ClientSession. Operations in the
+callback that run under that ClientSession can then extract the timeout from
+the session and apply it. To avoid confusing validation semantics, operations
+error if there is a timeout on the session but also an overridden timeout for
+the operation. It’s possible that the ability to communicate timeouts for a
+block of operations via a ClientSession is useful as a general purpose API,
+but we’ve decided to make it private until there are other known use cases.
+
+withTransaction refreshes the timeout for abortTransaction
+----------------------------------------------------------
+
+If the user-provided callback to ``withTransaction`` times out, it could
+leave a transaction running on the server. It’s imperative that drivers make
+an effort to abort the open transaction because failing to do so could result
+in the collections and databases affected by the transaction being locked for
+a long period of time, which could cause applications to stall. Because
+``timeoutMS`` has expired before drivers attempt to abort the transaction, we
+require drivers to refresh it and apply the original value to the execution
+of the ``abortTransaction`` operation. This can cause the entire
+``withTransaction`` call to take up to ``2*timeoutMS``, but it was decided
+that this risk is worthwhile given the importance of transaction cleanup.
```

Comment:
Is there any reason we can't also use `min(timeoutMS, connectTimeoutMS)` for `abortTransaction`, as we do for `killCursors` (see: "Cursor close() methods use connectTimeoutMS as timeoutMS")? Do we expect aborting a transaction to take a significant amount of time beyond network latency?


 aborting a transaction significantly different from killing a cursor

Reply:
I don't remember exactly why we decided to refresh for one and use `min(timeoutMS, connectTimeoutMS)` for the other. I'll reach out to someone from the server team to discuss this again and make sure the result of the discussion is documented this time around.

Reply:
Note: this is still an open question.

Reply:
I think we talked about this yesterday and concluded that it was important to give `abortTransaction` the best chance of succeeding to avoid blocking other operations. Should this be resolved?

Reply:
Yes, we decided to refresh `timeoutMS` for both `abortTransaction` and `killCursors`. Resolving.